File: src/main/java/day1java/extra/Applicative.java
package day1java.extra;

import java.util.function.Function;

public class Applicative<T> {

    public final T value;

    public Applicative(T value) {
        this.value = value;
    }

    <R> Applicative<R> applyOn(Applicative<Function<T, R>> other) {
        return new Applicative<>(other.value.apply(value));
    }
}

File: src/main/kotlin/extra/Applicative.kt
package extra

import java.util.function.Function


interface Applicative<A> {

    val value: A


}
//
//infix fun <T, R> Applicative<T>.applyOn(other: Applicative<(T) -> R>): Applicative<R> = Applicative.map(other.value(value))
//
//fun <A, B, R> Applicative<Pair<A, B>.map2(f: (Pair<A, B>) -> R): Applicative<R> = TODO()